### PR TITLE
APIs to enable custom tickers

### PR DESCRIPTION
**Pull request by felthy**
Potential solve for #39:

* New exports from `Theatre.core` in both bundles:
  - `disableDefaultTicker()`: disables the default core ticker, so you can schedule ticks yourself
  - `enableDefaultTicker()`: re-enables the default core ticker
  -  `ticker`: the core ticker, so you can call `ticker.tick(time)` from your scheduler
* Equivalents also added to `Theatre.studio` for the studio ticker
* Export `Ticker` from `Theatre.core`, and support supplying custom ticker to `getProject()` and `studio.initialize()`

### CODE REVIEW

**Comment by felthy** [comment]
Hi @AriaMinaei, thanks for reviewing this, and thanks for putting out such a great library!

I don’t think that’s a concern for this PR, for a few reasons:

* If I choose to use `disableDefaultTicker()`, then it becomes my responsibility to call `ticker.time()` or not. Consider how theatre.is could help me decide; it could expose an additional method (say, `ticker.required()`) for me to call. But if I’m calling that, theatre.js could just as well perform that check itself inside `ticker.time()`. Therefore the additional api is not required.
* If I’m using the default behavior, then theatre.js is responsible, so how theatre.js handles it internally has no bearing on this new API.
* In the latter case, theatre.js is correctly using `requestAnimationFrame()`, which allows the browser to throttle or halt callbacks when the tab is inactive, or for any other reason.

What do you think?

We’re already using this at Active Theory so we can schedule theatre.js’ updates within our renderer’s scheduler. We also use it to enable more advanced effects like blending between two positions on the same sequence.

**Review by donmccurdy — COMMENTED** [review]
Thanks for this PR! 🙏

Supporting user control of the animation loop would be helpful for us as well. I agree with @felthy that there are two modes to consider here:

- (A) default ticker is enabled, and can be naïve (simple rAF) or clever (pause on idle, etc.) or somewhere in between
- (B) default ticker is disabled, user is responsible for calling `tick()` and theatre.js updates sheets synchronously

From that perspective, do we need to pause-on-idle to be fully implemented first, or would it be enough to have a plan for what that API will eventually look like, and implement a way of disabling the default ticker compatible with that API? Happy to help with this, if it's not too early.

Also note that most browsers pause rAF loops automatically in important cases:

> requestAnimationFrame() calls are paused in most browsers when running in background tabs or hidden [<iframe>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe)s in order to improve performance and battery life. ~ [rAF docs](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame)

File: theatre/core/src/coreTicker.ts [hunk 1]
```diff
@@ -4,12 +4,21 @@ const coreTicker = new Ticker()
...
+}
+
+export function disableDefaultTicker() {
+  defaultTickerEnabled = false
+}
+
 const onAnimationFrame = (t: number) => {
   coreTicker.tick(t)
-  window.requestAnimationFrame(onAnimationFrame)
+  if (defaultTickerEnabled) window.requestAnimationFrame(onAnimationFrame)
```
[the remark this answers is not in the document]
Comment (felthy) [reply theatre/core/src/coreTicker.ts]: You're right, thanks for catching that @donmccurdy. Fix pushed.

**Comment by vezwork** [comment]
Update: we are actively discussing this internally. We are considering making fundamental changes to the ticker API, so we want to keep this PR open until we figure that out. 

Thanks @felthy for the great explanation and comments and thanks to both @felthy and @donmccurdy for the discussion.

**Comment by AndrewPrifer** [comment]
I have an alternative solution here: https://github.com/theatre-js/theatre/pull/230, curious about all your opinions @felthy @donmccurdy.

**Comment by felthy** [comment]
An important use case for us is flushing animation updates synchronously on demand (discussed in more detail [in this comment](https://github.com/theatre-js/theatre/pull/230#issuecomment-1163776003)).

When there are multiple active projects, flushing updates for a single project must not affect other projects.

To support this use case, I've exposed the `Ticker` constructor so that user code can create one Ticker per project:

```javascript
ticker = new Theatre.core.Ticker();
project = Theatre.core.getProject(name, { state: savedState, ticker });
// ... to flush animation updates for this project:
ticker.tick()
```

**Comment by AriaMinaei** [comment]
After a few iterations (#230 and #374) this PR is merged via d6498585295962c55a6d412e74ad7f65af86adb7. Will add to the release notes. Thank you @felthy!

**Pull request closed without merging**